src/builtins.cpp

```diff
@@ -1265,3 +1266,99 @@ void ispc::DefineStdlib(SymbolTable *symbolTable, llvm::LLVMContext *ctx, llvm::
+           retType->isSingleValueType() && retType->isDoubleTy());
+
+    std::string funcName = func.getName().str();
+    std::vector<llvm::Type *> ArgTy;
+#if ISPC_LLVM_VERSION == ISPC_LLVM_15_0
+    std::vector<SPIRV::PointerIndirectPair> PointerElementTys;
+#endif
+    // _DvWIDTH suffix is used in target file to differentiate scalar
+    // and vector versions of intrinsics. Here we remove this
+    // suffix and mangle the name.
+    size_t pos = funcName.find("_DvWIDTH");
+    bool isVaryingFunc = pos != std::string::npos;
+    if (isVaryingFunc) {
+        funcName.erase(pos, 8);
+    }
+    for (auto &arg : func.args()) {
+        ArgTy.push_back(arg.getType());
+        // In LLVM15 SPIR-V translator requires to pass pointer type information to mangleBuiltin
+        // https://github.com/KhronosGroup/SPIRV-LLVM-Translator/commit/0eb9a7d2937542e1f95a4e1f9aa9850e669dc45f
+        // It changes again in LLVM16 SPIR-V translator where TypedPointerType is required
+        // https://github.com/KhronosGroup/SPIRV-LLVM-Translator/commit/42cf770344bb8d0a32db1ec892bee63f43d793b1
+#if ISPC_LLVM_VERSION == ISPC_LLVM_15_0
+        SPIRV::PointerIndirectPair PtrElemTy;
+        if (arg.getType()->isPointerTy()) {
+            PtrElemTy.setPointer(isVaryingFunc ? LLVMTypes::DoubleVectorType : LLVMTypes::DoubleType);
+        }
+        PointerElementTys.push_back(PtrElemTy);
+#endif
+    }
+
+    mangleOpenClBuiltin(funcName, ArgTy,
+#if ISPC_LLVM_VERSION == ISPC_LLVM_15_0
+                        PointerElementTys,
+#endif
+                        mangledName);
+    return mangledName;
+}
+
+std::string ispc::manglePrintfOCLBuiltin(const llvm::Function &func) {
+    Assert(func.getName() == "__spirv_ocl_printf" && "wrong argument: ocl builtin is expected");
+    std::string mangledName;
+    mangleOpenClBuiltin(func.getName().str(), func.getArg(0)->getType(),
+#if ISPC_LLVM_VERSION == ISPC_LLVM_15_0
+                        // For spirv_ocl_printf builtin the argument is always i8*
+                        SPIRV::PointerIndirectPair(LLVMTypes::Int8Type),
+#endif
+                        mangledName);
+    return mangledName;
+}
+
+std::string ispc::mangleOCLBuiltin(const llvm::Function &func) {
+    Assert(func.getName().startswith("__spirv_ocl") && "wrong argument: ocl builtin is expected");
+    if (func.getName() == "__spirv_ocl_printf")
+        return manglePrintfOCLBuiltin(func);
+    return mangleMathOCLBuiltin(func);
+}
+
+std::string ispc::mangleSPIRVBuiltin(const llvm::Function &func) {
+    Assert(func.getName().startswith("__spirv_") && "wrong argument: spirv builtin is expected");
+    const std::string fnName = func.getName().str();
+    std::string mangledName;
+
+    if (func.getName().startswith("__spirv_SpecConstant")) {
+        constexpr std::size_t MAX_ARGS{1024UL};
+        std::size_t i = 0;
+        llvm::SmallVector<llvm::Type *, MAX_ARGS> tyArgs;
```

Comment:
Why not to simply use `std::vector<llvm::Type *>tyArgs` instead of `llvm::SmallVector<llvm::Type *, MAX_ARGS>`?